README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# recount <img src="man/figures/logo.png" align="right" width="400px" />

<!-- badges: start -->

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![Bioc release
status](http://www.bioconductor.org/shields/build/release/bioc/recount.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/recount)
[![Bioc devel
status](http://www.bioconductor.org/shields/build/devel/bioc/recount.svg)](https://bioconductor.org/checkResults/devel/bioc-LATEST/recount)
[![Bioc downloads
rank](https://bioconductor.org/shields/downloads/release/recount.svg)](http://bioconductor.org/packages/stats/bioc/recount/)
[![Bioc
support](https://bioconductor.org/shields/posts/recount.svg)](https://support.bioconductor.org/tag/recount)
[![Bioc
history](https://bioconductor.org/shields/years-in-bioc/recount.svg)](https://bioconductor.org/packages/release/bioc/html/recount.html#since)
[![Bioc last
commit](https://bioconductor.org/shields/lastcommit/devel/bioc/recount.svg)](http://bioconductor.org/checkResults/devel/bioc-LATEST/recount/)
[![Bioc
dependencies](https://bioconductor.org/shields/dependencies/release/recount.svg)](https://bioconductor.org/packages/release/bioc/html/recount.html#since)
[![Codecov test
coverage](https://codecov.io/gh/leekgroup/recount/branch/devel/graph/badge.svg)](https://codecov.io/gh/leekgroup/recount?branch=devel)
[![R build
status](https://github.com/leekgroup/recount/actions/workflows/check-bioc.yml/badge.svg)](https://github.com/leekgroup/recount/actions/workflows/check-bioc.yml)
[![GitHub
issues](https://img.shields.io/github/issues/leekgroup/recount)](https://github.com/leekgroup/recount/issues)
[![GitHub
pulls](https://img.shields.io/github/issues-pr/leekgroup/recount)](https://github.com/leekgroup/recount/pulls)
<!-- badges: end -->

Explore and download data from the recount project available at the
[recount2 website](https://jhubiostatistics.shinyapps.io/recount/).
Using the `recount` package you can download
*RangedSummarizedExperiment* objects at the gene, exon or exon-exon
junctions level, the raw counts, the phenotype metadata used, the urls
to the sample coverage bigWig files or the mean coverage bigWig file for
a particular study. The *RangedSummarizedExperiment* objects can be used
by different packages for performing differential expression analysis.
Using [derfinder](http://bioconductor.org/packages/derfinder) you can
perform annotation-agnostic differential expression analyses with the
data from the recount project.

## Documentation

For more information about `recount` check the vignettes [through
Bioconductor](http://bioconductor.org/packages/recount) or at the
[documentation website](http://leekgroup.github.io/recount).

## Installation instructions

Get the latest stable `R` release from
[CRAN](http://cran.r-project.org/). Then install `recount` from
[Bioconductor](http://bioconductor.org/) using the following code:

``` r
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("recount")
```

## Citation

Below is the citation output from using `citation('recount')` in R.
Please run this yourself to check for any updates on how to cite
**recount**.

``` r
print(citation("recount"), bibtex = TRUE)
#> To cite package 'recount' in publications use:
#> 
#>   Collado-Torres L, Nellore A, Kammers K, Ellis SE, Taub MA, Hansen KD,
#>   Jaffe AE, Langmead B, Leek JT (2017). "Reproducible RNA-seq analysis
#>   using recount2." _Nature Biotechnology_. doi:10.1038/nbt.3838
#>   <https://doi.org/10.1038/nbt.3838>,
#>   <http://www.nature.com/nbt/journal/v35/n4/full/nbt.3838.html>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {Reproducible RNA-seq analysis using recount2},
#>     author = {Leonardo Collado-Torres and Abhinav Nellore and Kai Kammers and Shannon E. Ellis and Margaret A. Taub and Kasper D. Hansen and Andrew E. Jaffe and Ben Langmead and Jeffrey T. Leek},
#>     year = {2017},
#>     journal = {Nature Biotechnology},
#>     doi = {10.1038/nbt.3838},
#>     url = {http://www.nature.com/nbt/journal/v35/n4/full/nbt.3838.html},
#>   }
#> 
#>   Collado-Torres L, Nellore A, Jaffe AE (2017). "recount workflow:
#>   Accessing over 70,000 human RNA-seq samples with Bioconductor
#>   [version 1; referees: 1 approved, 2 approved with reservations]."
#>   _F1000Research_. doi:10.12688/f1000research.12223.1
#>   <https://doi.org/10.12688/f1000research.12223.1>,
#>   <https://f1000research.com/articles/6-1558/v1>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {recount workflow: Accessing over 70,000 human RNA-seq samples with Bioconductor [version 1; referees: 1 approved, 2 approved with reservations]},
#>     author = {Leonardo Collado-Torres and Abhinav Nellore and Andrew E. Jaffe},
#>     year = {2017},
#>     journal = {F1000Research},
#>     doi = {10.12688/f1000research.12223.1},
#>     url = {https://f1000research.com/articles/6-1558/v1},
#>   }
#> 
#>   Ellis SE, Collado-Torres L, Jaffe AE, Leek JT (2018). "Improving the
#>   value of public RNA-seq expression data by phenotype prediction."
#>   _Nucl. Acids Res._. doi:10.1093/nar/gky102
#>   <https://doi.org/10.1093/nar/gky102>,
#>   <https://doi.org/10.1093/nar/gky102>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {Improving the value of public RNA-seq expression data by phenotype prediction},
#>     author = {Shannon E. Ellis and Leonardo Collado-Torres and Andrew E. Jaffe and Jeffrey T. Leek},
#>     year = {2018},
#>     journal = {Nucl. Acids Res.},
#>     doi = {10.1093/nar/gky102},
#>     url = {https://doi.org/10.1093/nar/gky102},
#>   }
#> 
#>   Collado-Torres L, Nellore A, Kammers K, Ellis SE, Taub MA, Hansen KD,
#>   Jaffe AE, Langmead B, Leek JT (2023). _Explore and download data from
#>   the recount project_. doi:10.18129/B9.bioc.recount
#>   <https://doi.org/10.18129/B9.bioc.recount>,
#>   https://github.com/leekgroup/recount - R package version 1.27.0,
#>   <http://www.bioconductor.org/packages/recount>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Manual{,
#>     title = {Explore and download data from the recount project},
#>     author = {Leonardo Collado-Torres and Abhinav Nellore and Kai Kammers and Shannon E. Ellis and Margaret A. Taub and Kasper D. Hansen and Andrew E. Jaffe and Ben Langmead and Jeffrey T. Leek},
#>     year = {2023},
#>     url = {http://www.bioconductor.org/packages/recount},
#>     note = {https://github.com/leekgroup/recount - R package version 1.27.0},
#>     doi = {10.18129/B9.bioc.recount},
#>   }
#> 
#>   Frazee AC, Langmead B, Leek JT (2011). "ReCount: A multi-experiment
#>   resource of analysis-ready RNA-seq gene count datasets." _BMC
#>   Bioinformatics_. doi:10.1186/1471-2105-12-449
#>   <https://doi.org/10.1186/1471-2105-12-449>,
#>   <https://doi.org/10.1186/1471-2105-12-449>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {ReCount: A multi-experiment resource of analysis-ready RNA-seq gene count datasets},
#>     author = {Alyssa C. Frazee and Ben Langmead and Jeffrey T. Leek},
#>     year = {2011},
#>     journal = {BMC Bioinformatics},
#>     doi = {10.1186/1471-2105-12-449},
#>     url = {https://doi.org/10.1186/1471-2105-12-449},
#>   }
#> 
#>   Razmara A, Ellis SE, Sokolowski DJ, Davis S, Wilson MD, Leek JT,
#>   Jaffe AE, Collado-Torres L (2019). "recount-brain: a curated
#>   repository of human brain RNA-seq datasets metadata." _bioRxiv_.
#>   doi:10.1101/618025 <https://doi.org/10.1101/618025>,
#>   <https://doi.org/10.1101/618025>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {recount-brain: a curated repository of human brain RNA-seq datasets metadata},
#>     author = {Ashkaun Razmara and Shannon E. Ellis and Dustin J. Sokolowski and Sean Davis and Michael D. Wilson and Jeffrey T. Leek and Andrew E. Jaffe and Leonardo Collado-Torres},
#>     year = {2019},
#>     journal = {bioRxiv},
#>     doi = {10.1101/618025},
#>     url = {https://doi.org/10.1101/618025},
#>   }
#> 
#>   Imada E, Sanchez DF, Collado-Torres L, Wilks C, Matam T, Dinalankara
#>   W, Stupnikov A, Lobo-Pereira F, Yip C, Yasuzawa K, Kondo N, Itoh M,
#>   Suzuki H, Kasukawa T, Hon CC, de Hoon MJ, Shin JW, Carninci P, Jaffe
#>   AE, Leek JT, Favorov A, Franco GR, Langmead B, Marchionni L (2020).
#>   "Recounting the FANTOM CAGE–Associated Transcriptome." _Genome
#>   Research_. doi:10.1101/gr.254656.119
#>   <https://doi.org/10.1101/gr.254656.119>,
#>   <https://doi.org/10.1101/gr.254656.119>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {Recounting the FANTOM CAGE–Associated Transcriptome},
#>     author = {Eddie-Luidy Imada and Diego Fernando Sanchez and Leonardo Collado-Torres and Christopher Wilks and Tejasvi Matam and Wikum Dinalankara and Aleksey Stupnikov and Francisco Lobo-Pereira and Chi-Wai Yip and Kayoko Yasuzawa and Naoto Kondo and Masayoshi Itoh and Harukazu Suzuki and Takeya Kasukawa and Chung Chau Hon and Michiel JL {de Hoon} and Jay W Shin and Piero Carninci and Andrew E. Jaffe and Jeffrey T. Leek and Alexander Favorov and Glória R Franco and Ben Langmead and Luigi Marchionni},
#>     year = {2020},
#>     journal = {Genome Research},
#>     doi = {10.1101/gr.254656.119},
#>     url = {https://doi.org/10.1101/gr.254656.119},
#>   }
```

Please note that the `recount` was only made possible thanks to many
other R and bioinformatics software authors, which are cited either in
the vignettes and/or the paper(s) describing this package.

## Code of Conduct

Please note that the recount project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## Development tools

- Continuous code testing is possible thanks to [GitHub
  actions](https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/)
  through *[usethis](https://CRAN.R-project.org/package=usethis)*,
  *[remotes](https://CRAN.R-project.org/package=remotes)*,
  *[sysreqs](https://github.com/r-hub/sysreqs)* and
  *[rcmdcheck](https://CRAN.R-project.org/package=rcmdcheck)* customized
  to use [Bioconductor’s docker
  containers](https://www.bioconductor.org/help/docker/) and
  *[BiocCheck](https://bioconductor.org/packages/3.17/BiocCheck)*.
- Code coverage assessment is possible thanks to
  [codecov](https://codecov.io/gh) and
  *[covr](https://CRAN.R-project.org/package=covr)*.
- The [documentation website](http://leekgroup.github.io/recount) is
  automatically updated thanks to
  *[pkgdown](https://CRAN.R-project.org/package=pkgdown)*.
- The code is styled automatically thanks to
  *[styler](https://CRAN.R-project.org/package=styler)*.
- The documentation is formatted thanks to
  *[devtools](https://CRAN.R-project.org/package=devtools)* and
  *[roxygen2](https://CRAN.R-project.org/package=roxygen2)*.

For more details, check the `dev` directory.

This package was developed using
*[biocthis](https://bioconductor.org/packages/3.17/biocthis)*.

## Teams involved

- [Jeff Leek’s lab at JHBSPH Biostatistics
  Department](http://jtleek.com/),
- [Ben Langmead’s lab at JHU Computer
  Science](http://www.langmead-lab.org/),
- [Kasper Daniel Hansen’s lab at JHBSPH Biostatistics
  Department](https://www.hansenlab.org/),
- [Leonardo Collado-Torres](http://lcolladotor.github.io/) and
  [Andrew E. Jaffe](http://aejaffe.com/) from
  [LIBD](https://www.libd.org/),
- [Abhinav Nellore’s lab at OHSU](http://nellore.bio/),
- Data hosted by [SciServer at JHU](https://www.sciserver.org/).

|                                                                                                                                                                               |                                                                                                              |                                                                                                                                                                         |                                                                                                                                                   |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| <a href="http://www.langmead-lab.org/"><img src="http://www.langmead-lab.org/wp-content/uploads/2014/01/Screen-Shot-2014-02-02-at-5.20.13-PM-1024x199.png" width="250px"></a> | <a href="https://www.libd.org/"><img src="http://lcolladotor.github.io/img/LIBD_logo.jpg" width="250px"></a> | <a href="http://nellore.bio/"><img src="https://seekvectorlogo.net/wp-content/uploads/2018/08/oregon-health-science-university-ohsu-vector-logo.png" width="250px"></a> | <a href="https://www.sciserver.org/"><img src="https://skyserver.sdss.org/dr14/en/images/sciserver_logo_inverted_vertical.png" width="250px"></a> |

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-78422749-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-78422749-1');
</script>


NEWS
CHANGES IN VERSION 1.8.2
------------------------

BUG FIXES

    o Added the argument `async` to snaptron_query() which
        can be set to FALSE to address the issue reported
        at https://github.com/ChristopherWilks/snaptron/issues/11


CHANGES IN VERSION 1.7.7
------------------------

BUG FIXES

    o Updated reproduce_ranges() to match the URL change in
        Gencode from
        ftp://ftp.sanger.ac.uk to ftp://ftp.ebi.ac.uk


CHANGES IN VERSION 1.7.5
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o add_metadata() can now download the recount_brain_v2 data.


CHANGES IN VERSION 1.7.4
------------------------

BUG FIXES

    o Fix a NOTE about RefManageR.


CHANGES IN VERSION 1.7.3
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Use BiocManager

CHANGES IN VERSION 1.7.2
------------------------

BUG FIXES

    o Fix a unit test.


CHANGES IN VERSION 1.7.1
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o rse_tx URLs now point to v2 to reflect recent changes by Fu et al.


CHANGES IN VERSION 1.5.11
-------------------------

BUG FIXES

    o Change some examples to dontrun and improve the code that cleans up after
        the tests. This should reduce the size of files left in tmp although
        they didn't seem too big to begin with.


CHANGES IN VERSION 1.5.9
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o The functions add_metadata() and add_predictions() now return the
        sample metadata or predictions when the 'rse' argument is missing.

CHANGES IN VERSION 1.5.6
------------------------

NEW FEATURES

    o Added the function add_metadata() which can be used to append curated
        metadata to a recount rse object. Currently, add_metadata() only
        supports the recount_brain_v1 data available at
        http://lieberinstitute.github.io/recount-brain/
        and to be further described in Razmara et al, in prep, 2018.

CHANGES IN VERSION 1.5.5
------------------------

BUG FIXES

    o Fix doc link in geo_characteristics() which affected the Windows build
        machines.

CHANGES IN VERSION 1.5.4
------------------------

BUG FIXES

    o Fix a unit test for download_study(), add another test for the versions,
        and fix a NOTE in R CMD check.

CHANGES IN VERSION 1.5.3
------------------------

NEW FEATURES

    o download_study() can now download the transcript counts (rse_tx.RData)
        files. The transcript estimation is described in Fu et al, 2018.
        
SIGNIFICANT USER-VISIBLE CHANGES

    o download_study() now has a version parameter (defaults to 2). This
        argument controls which version of the files to download based on the
        change on how exons were defined. Version 1 are reduced exons while
        version 2 are disjoint exons as described in further detail in the
        documentation tab of the recount website 
        https://jhubiostatistics.shinyapps.io/recount/.
    o recount_url and the example rse_gene_SRP009615 have been updated to match
        the changes in version 2.


CHANGES IN VERSION 1.3.13
-------------------------

BUG FIXES

    o Changed reproduce_ranges() since disjoint exons are more useful than
        reduced exons for downstream analyses.


CHANGES IN VERSION 1.3.12
-------------------------

NEW FEATURES

    o Added the function read_counts().


CHANGES IN VERSION 1.3.9
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Added citations for 
        https://www.biorxiv.org/content/early/2017/06/03/145656 and
        https://f1000research.com/articles/6-1558/v1 as well as mentions to
        them in the vignette.


CHANGES IN VERSION 1.3.7
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o add_predictions() was bumped to version 0.0.05


CHANGES IN VERSION 1.3.5
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Vignette now uses the new BiocStyle::html_document that was recently
        released.


CHANGES IN VERSION 1.3.2
------------------------

NEW FEATURES

    o coverage_matrix() now has two new arguments: scale and round. Use
        scale = FALSE to get raw coverage counts, which you can then scale with
        scale_counts(). scale is set to TRUE by default, so the counts are
        scaled to a library size of 40 million reads. round is set to FALSE by
        default, but can be set to TRUE if you want to get integer counts, just
        as in the default of scale_counts().

CHANGES IN VERSION 1.3.1
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Changed the default version argument of add_predictions() to 'latest'.
        Internally, that's still 0.0.03.


CHANGES IN VERSION 1.1.27
-------------------------

NEW FEATURES

    o Added the add_predictions() function which appends the predicted
        phenotypes to a RSE object downloaded with recount. The phenotypes
        were predicted by Shannon Ellis et al, 2017 (citation coming up soon!).

CHANGES IN VERSION 1.1.26
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Changed the citation now that the recount2 paper has been published at
        http://www.nature.com/nbt/journal/v35/n4/full/nbt.3838.html.


CHANGES IN VERSION 1.1.25
-------------------------

NEW FEATURES

    o Added the function getRPKM() which can be used with
        RangedSummarizedExperiment objects from recount and from other sources.

CHANGES IN VERSION 1.1.24
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o recount_url now includes the URLs for the GTEx bigWig files.

CHANGES IN VERSION 1.1.19
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o coverage_matrix() now returns a RangedSummarizedExperiment object. This
        matches the behavior of recount.bwtool::coverage_matrix_bwtool() and
        is more consistent with the use of RSE objects in recount.

CHANGES IN VERSION 1.1.18
-------------------------

BUG FIXES


tests/testthat/test-data.R
context('Download data, scale, ER-level')

library('SummarizedExperiment')

## Download the example data included in the package
url <- download_study('SRP009615', outdir = file.path(tempdir(), 'SRP009615'))

## Load the data
load(file.path(tempdir(), 'SRP009615', 'rse_gene.Rdata'))

## Compare the data
test_that('Example RSE', {
    expect_equivalent(rse_gene, rse_gene_SRP009615)
})

## Temporary download URLs
test_that('Download URLs', {
    expect_equal(download_study('DRP000366', type = 'mean', download = FALSE),
        'http://duffel.rail.bio/recount/DRP000366/bw/mean_DRP000366.bw')
    expect_equal(download_study('DRP000366', type = 'samples',
        download = FALSE),
        'http://duffel.rail.bio/recount/DRP000366/bw/DRR000897.bw')
})

## Test downloading a small project entirely
tmpdir <- file.path(tempdir(), 'SRP002001')
urls <- download_study('SRP002001', type = 'all', outdir = tmpdir)
expected_urls <- paste0('http://duffel.rail.bio/recount/SRP002001/',
    c('rse_gene.Rdata', 'rse_exon.Rdata', 'rse_jx.Rdata', 'counts_gene.tsv.gz',
        'counts_exon.tsv.gz', 'counts_jx.tsv.gz', 
        'SRP002001.tsv', 'files_info.tsv',
        'bw/SRR036661.bw', 'bw/mean_SRP002001.bw'))
names(expected_urls) <- c('rse-gene', 'rse-exon', 'rse-jx', 
    'counts-gene', 'counts-exon', 'counts-jx',
    'phenotype', 'files-info', 'samples', 'mean')

## Compute md5sum locally
localfiles <- c(list.files(tmpdir, '[.]', full.names = TRUE),
    dir(file.path(tmpdir, 'bw'), full.names = TRUE))
names(localfiles) <- c(list.files(tmpdir, '[.]'), dir(file.path(tmpdir, 'bw')))

library('tools')
md5 <- sapply(localfiles, md5sum)
names(md5) <- names(localfiles)
md5 <- md5[-which(names(md5) == 'files_info.tsv')]

## Get original md5sum
fileinfo <- read.table(file.path(tmpdir, 'files_info.tsv'), header = TRUE,
    stringsAsFactors = FALSE)

test_that('Project SRP002001', {
    expect_equal(urls, expected_urls)
    expect_equivalent(fileinfo$md5sum[match(names(md5), fileinfo$file)], md5)
})




scaleFac <- scale_counts(rse_gene_SRP009615, factor_only = TRUE)
scaleFac_mapped <- scale_counts(rse_gene_SRP009615, by = 'mapped_reads', 
    factor_only = TRUE)
rse <- scale_counts(rse_gene_SRP009615, round = FALSE)

test_that('Scaling', {
    expect_equal(round(head(scaleFac), 2), c('SRR387777' = 0.04,
        'SRR387778' = 0.03, 'SRR387779' = 0.03, 'SRR387780' = 0.04,
        'SRR389077' = 0.04, 'SRR389078' = 0.04))
    expect_gt(scaleFac_mapped[1], scaleFac[1])
    expect_gt(scaleFac_mapped[2], scaleFac[2])
    expect_gt(scaleFac_mapped[3], scaleFac[3])
    expect_equal(assay(rse, 1) / matrix(rep(scaleFac, each = 58037),
        ncol = 12), assay(rse_gene_SRP009615, 1))
})

if(.Platform$OS.type != 'windows') {
    regions <- expressed_regions('SRP002001', 'chrY', cutoff = 5)
    ## Artificially remove the mean coverage file so that the file will have to
    ## get downloaded on the first test, then it'll be present for the second
    ## test
    unlink(localfiles['mean_SRP002001.bw'])

    test_that('Expressed regions', {
        expect_equal(regions,
            expressed_regions('SRP002001', 'chrY', cutoff = 5, outdir = tmpdir))
        expect_equal(regions,
            expressed_regions('SRP002001', 'chrY', cutoff = 5, outdir = tmpdir))
    })


    coverageMatrix <- coverage_matrix('SRP002001', 'chrY', regions)
    ## Same for the phenotype data and the sample bigwig file
    unlink(localfiles['SRP002001.tsv'])
    unlink(localfiles['SRR036661.bw'])

    test_that('Coverage matrix', {
        expect_equal(coverageMatrix,
            coverage_matrix('SRP002001', 'chrY', regions, outdir = tmpdir))
        expect_equal(coverageMatrix,
            coverage_matrix('SRP002001', 'chrY', regions, outdir = tmpdir,
            chunksize = 500))
    })
}

metadata <- all_metadata()
test_that('All metadata', {
    expect_equal(nrow(metadata), 50099)
})

phenoFile <- download_study(project = 'SRP012289', type = 'phenotype',
    download = FALSE)
pheno <- read.table(phenoFile, header = TRUE, stringsAsFactors = FALSE,
    sep = '\t')
test_that('Correct phenotype information', {
    expect_equal(pheno$auc, 159080954)
})


## Test Snaptron
library('GenomicRanges')
junctions <- GRanges(seqnames = 'chr2', IRanges(
    start = c(28971710:28971712, 29555081:29555083, 29754982:29754984),
    end = c(29462417:29462419, 29923338:29923340, 29917714:29917716)))
    
junctions_v2 <- GRanges(seqnames = 'chr2', IRanges(
    start = 29532116:29532118, end = 29694848:29694850)) 

snap <- snaptron_query(junctions)
snap_v2 <- snaptron_query(junctions_v2, version = 'srav2')
snap_gtex <- snaptron_query(junctions_v2, version = 'gtex')
snap_tcga <- snaptron_query(junctions_v2, version = 'tcga')

test_that('Snaptron', {
    expect_equal(length(snap), 3)
    expect_equal(ncol(mcols(snap)), 14)
    expect_equal(snap$left_annotated[[1]], as.character(NA))
    expect_equal(snaptron_query(junctions[1], verbose = FALSE), NULL)
    expect_equal(is(snap_v2$annotated, 'CompressedCharacterList'), TRUE)
    expect_equal(snaptron_query(junctions_v2, verbose = FALSE), NULL)
    expect_equal(snap_gtex$type == 'GTEx:I', TRUE)
    expect_equal(snap_tcga$type == 'TCGA:I', TRUE)
})

## Weird pheno files
## First 2 are separate from the rest
projects <- c('SRP036843', 'SRP029334', 'SRP050563', 'SRP055438', 'SRP055749',
    'SRP058120', 'SRP005342', 'SRP007508', 'SRP015668')
sapply(projects, download_study, type = 'phenotype')
phenoFiles <- sapply(projects, function(x) { file.path(x, paste0(x, '.tsv')) })
pheno <- mapply(recount:::.read_pheno, phenoFiles, projects, SIMPLIFY = FALSE)

pheno_tcga <- recount:::.read_pheno('doesntexist', 'TCGA')
test_that('Weird pheno files', {
    expect_equivalent(sapply(pheno, nrow), c(3, 5, 4, 33, 16, 30, 12, 5, 33))
    expect_equal(dim(pheno_tcga), c(11284, 864))
})


R/snaptron_query.R
#' Query Snaptron to get data from exon-exon junctions present in Intropolis
#'
#' This function uses the Snaptron API to query specific exon-exon junctions
#' that are available via Intropolis as described in the vignette.
#'
#' @param junctions A \link[GenomicRanges]{GRanges-class} object with the
#' exon-exon junctions of interest. The chromosome names should be in UCSC
#' format, such as 'chr1'. The strand information is ignored in the query.
#' @param version Either \code{1} or \code{2}. Version 1 of Intropolis has the
#' exon-exon junctions from about 20 thousand RNA-seq samples in hg19 
#' coordinates. Version 2 has the data from about 50 thousand RNA-seq samples #' aligned to hg38.
#' @param verbose If \code{TRUE} status updates will be printed.
#'
#' @return A \link[GenomicRanges]{GRanges-class} object with the results from
#' the Snaptron query. For information on the different columns please see
#' \url{http://snaptron.cs.jhu.edu/snaptron/docs/}.
#'
#' @references Please cite \url{http://snaptron.cs.jhu.edu/snaptron/docs/}
#' if you use this function as Snaptron is a separate project from recount.
#' Thank you!
#'
#' @author Leonardo Collado-Torres
#' @export
#'
#' @import GenomicRanges
#' @import RCurl
#'
#' @examples
#' 
#' library('GenomicRanges')
#' ## Define some exon-exon junctions (hg19 coordinates)
#' junctions <- GRanges(seqnames = 'chr2', IRanges(
#'     start = c(28971710:28971712, 29555081:29555083, 29754982:29754984),
#'     end = c(29462417:29462419, 29923338:29923340, 29917714:29917716)))
#'
#' ## Check against Snaptron version 1 (hg19 coordinates)
#' snaptron_query(junctions)
#'
#' ## Check another set of junctions against version 2 (more data, hg38 
#' ## coordinates)
#' junctions_v2 <- GRanges(seqnames = 'chr2', IRanges(
#'     start = 29532116:29532118, end = 29694848:29694850))
#' snaptron_query(junctions_v2, version = 2) 
#'

snaptron_query <- function(junctions, version = 1, verbose = TRUE) {
    ## Check input
    stopifnot(is(junctions, 'GRanges'))
    stopifnot(all(grepl('chr', seqlevels(junctions))))
    stopifnot(version %in% c(1, 2))
    ver <- ifelse(version == 1, 'srav1', 'srav2')
    
    
    ## Build query URLs
    urls <- paste0('http://stingray.cs.jhu.edu:8090/', ver,
        '/snaptron?regions=', seqnames(junctions), ':', start(junctions), '-',
        end(junctions), '&exact=1&header=0')
    
    ## Get results
    if(verbose) message(paste(Sys.time(), 'querying Snaptron'))
    query_res <- getURL(urls)
    
    ## Split by line
    if(verbose) message(paste(Sys.time(), 'processing results'))
    query_split <- strsplit(query_res, '\n')
    names(query_split) <- NULL
    
    ## Are there any valid ones?
    valid <- which(elementNROWS(query_split) > 0)
    if(length(valid) == 0) {
        message(paste(Sys.time(),
            'found no exon-exon junctions in Intropolis version', version,
            'matching your query: this version uses',
            ifelse(version == 1, 'hg19', 'hg38'), 'coordinates.'))
        return(NULL)
    }
    
    ## Extract information
    if(verbose) message(paste(Sys.time(), 'extracting information'))
    info <- lapply(query_split[valid], function(jxs) { 
        matrix(strsplit(jxs, '\t')[[1]], ncol = 19)
    })
    
    ## Build result
    res <- do.call(rbind, info)
    colnames(res) <- c('type', 'snaptron_id', 'chromosome', 'start', 'end',
        'length', 'strand', 'annotated', 'left_motif', 'right_motif',
        'left_annotated', 'right_annotated', 'samples',
        'read_coverage_by_sample', 'samples_count', 'coverage_sum',
        'coverage_avg', 'coverage_median', 'source_dataset_id')
    
    ## Helper function for some special variables
    to_chr_list <- function(x) {
        r <- strsplit(x, ',')
        i <- which(sapply(r, function(y) { y[[1]] == '0' }))
        if(length(i) > 0) r[i] <- NA
        return(CharacterList(r))
    }
    
    ## Build into a GRanges object
    result <- GRanges(seqnames = res[, 'chromosome'], 
        IRanges(as.numeric(res[, 'start']), as.numeric(res[, 'end'])),
        strand = res[, 'strand'])
    
    ## Add other variables
    result$type <- as.factor(res[, 'type'])
    result$snaptron_id <- as.integer(res[, 'snaptron_id'])
    result$annotated <- to_chr_list(res[, 'annotated'])
    result$left_motif <- res[, 'left_motif']
    result$right_motif <- res[, 'right_motif']
    result$left_annotated <- to_chr_list(res[, 'left_annotated'])
    result$right_annotated <- to_chr_list(res[, 'right_annotated'])
    result$samples <- IntegerList(strsplit(res[, 'samples'], ','))
    result$read_coverage_by_sample <- IntegerList(strsplit(res[,
        'read_coverage_by_sample'], ','))
    result$samples_count <- as.integer(res[, 'samples_count'])
    result$coverage_sum <- as.integer(res[, 'coverage_sum'])
    result$coverage_avg <- as.numeric(res[, 'coverage_avg'])
    result$coverage_median <- as.numeric(res[, 'coverage_median'])
    result$source_dataset_id <- as.integer(res[, 'source_dataset_id'])
    
    ## Finish
    return(result)
}
